tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs

```diff
@@ -276,6 +276,45 @@ public void Dispose_should_raise_the_correct_events()
             act.ShouldThrow<InvalidOperationException>();
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void ReceiveMessage_should_throw_error_when_message_is_an_invalid_size(
+            [Values(-1, 16000005)]
+            int length,
+            [Values(false, true)]
+            bool async)
+        {
+            using (var stream = new BlockingMemoryStream())
+            {
+                var bytes = BitConverter.GetBytes(length);
+                stream.Write(bytes, 0, bytes.Length);
+                stream.Seek(0, SeekOrigin.Begin);
+                var encoderSelector = new ReplyMessageEncoderSelector<BsonDocument>(BsonDocumentSerializer.Instance);
+
+                Exception invalidSizeException;
+                if (async)
+                {
+                    _mockStreamFactory.Setup(f => f.CreateStreamAsync(_endPoint, CancellationToken.None))
+                        .Returns(Task.FromResult<Stream>(stream));
+                    _subject.OpenAsync(CancellationToken.None).GetAwaiter().GetResult();
+                    invalidSizeException = Record.Exception(() => _subject
+                        .ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None)
+                        .GetAwaiter().GetResult());
+                }
+                else
+                {
+                    _mockStreamFactory.Setup(f => f.CreateStream(_endPoint, CancellationToken.None))
+                        .Returns(stream);
+                    _subject.Open(CancellationToken.None);
+
+                    invalidSizeException = Record.Exception( () =>  _subject.ReceiveMessage(10, encoderSelector, _messageEncoderSettings, CancellationToken.None));
+                }
+
+                Assert.NotNull(invalidSizeException);
```

Comment:
Sorry for the style confusion Katie! I believe that this line and the line below it should probably use Fluent Assertions ([example](https://github.com/mongodb/mongo-csharp-driver/blob/48e7a08bb4069cdbcdfb4f12459db7001a842abd/tests/MongoDB.Driver.Core.Tests/Core/Operations/EndTransactionOperationTests.cs#L43))